node-admin/src/main/java/com/yahoo/vespa/hosted/node/admin/maintenance/servicedump/VespaServiceDumperImpl.java

```diff
@@ -0,0 +1,163 @@
+        } catch (Exception e) {
+            handleFailure(context, request, startedAt, e, e.getMessage());
+        }
+    }
+
+    private List<SyncFileInfo> dumpFiles(Path directoryOnHost, URI destination, Instant expiry) {
+        return FileFinder.files(directoryOnHost).stream()
+                .flatMap(file -> SyncFileInfo.forServiceDump(destination, file.path(), expiry).stream())
+                .collect(Collectors.toList());
+    }
+
+    private void logFilesToUpload(NodeAgentContext context, List<SyncFileInfo> files) {
+        if (log.isLoggable(Level.FINE)) {
+            String message = files.stream()
+                    .map(file -> file.source().toString())
+                    .collect(Collectors.joining());
+            context.log(log, Level.FINE, message);
+        }
+    }
+
+    private static Instant expiry(Instant startedAt, ServiceDumpReport request) {
+        return request.expiry() != null
+                ? Instant.ofEpochMilli(request.expiry())
+                : startedAt.plus(7, ChronoUnit.DAYS);
+    }
+
+    private void handleFailure(NodeAgentContext context, ServiceDumpReport request, Instant startedAt,
+                               Exception failure, String message) {
+        if (failure != null) {
+            context.log(log, Level.WARNING, message, failure);
+        } else {
+            context.log(log, Level.WARNING, message);
+        }
+        ServiceDumpReport report = createErrorReport(request, startedAt, message);
+        storeReport(context, report);
+    }
+
+    private void storeReport(NodeAgentContext context, ServiceDumpReport report) {
+        NodeAttributes nodeAttributes = new NodeAttributes();
+        nodeAttributes.withReport(ServiceDumpReport.REPORT_ID, report.toJsonNode());
+        nodeRepository.updateNodeAttributes(context.hostname().value(), nodeAttributes);
+    }
+
+    private static ServiceDumpReport createStartedReport(ServiceDumpReport request, Instant startedAt) {
+        return new ServiceDumpReport(
+                request.requestedAt(), startedAt.toEpochMilli(), null, null, null, request.configId(),
+                request.expiry(), null);
+    }
+
+    private static ServiceDumpReport createSuccessReport(ServiceDumpReport request, Instant startedAt, URI location) {
+        return new ServiceDumpReport(
+                request.requestedAt(), startedAt.toEpochMilli(), Instant.now().toEpochMilli(), null,
+                location.toString(), request.configId(), request.expiry(), null);
+    }
+
+    private static ServiceDumpReport createErrorReport(ServiceDumpReport request, Instant startedAt, String message) {
+        return new ServiceDumpReport(
+                request.requestedAt(), startedAt.toEpochMilli(), null, Instant.now().toEpochMilli(), null,
+                request.configId(), request.expiry(), message);
+    }
+
+    private static String createDumpId(ServiceDumpReport report) {
+        String sanitizedConfigId = report.configId()
```

Comment:
Make sure all characters of the config ID is sanitized?  This ends up as part of URI, so it sounds like it ought to be urlencoded at some point, unless the characters can be guaranteed to be valid pathname chars.